### ISSUE TITLE
Unexpected "read before assignment to local variable 'foo'"

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#4407]
Compiling the following code results in an error with message `read before assignment to local variable 'foo'`:

```crystal
module Test
  class_property foo = 1
end

def foo
  Test.foo
end

foo
```

It might have something to do with the `class_property` macro polluting global scope. If the property is defined like this, everything works as expected:
```
  class_property foo
  @@foo = 1
```

Tested on v0.22.0 [Playground Link](https://play.crystal-lang.org/#/r/20ns)

**Comment by straight-shoota** [comment]
This works:

```cr
module Test
  {% begin %}
  class_property {{"foo = 1".id}}
  {% end %}
end

def foo
  Test.foo
end

foo
```

So maybe the parser somehow assumes `foo = 1` declares a local variable in top level scope - which is already weird - and for some reason does not even assign a value.

**Comment by asterite** [comment]
Yes, that's the reason. And I don't find it weird. Or at least I have no idea how to fix it :-P

**Comment by straight-shoota** [comment]
Btw, is there any particular reason to share local variables between local scope and type scope?
```cr
module Test
  foo = 1
end

p! foo
```

**Comment by straight-shoota** [comment]
That doesn't even work in Ruby 😛

**Comment by asterite** [comment]
I mean, this works:

```
puts a = 1
puts a
```

The scope issue is another issue: at the time we implemented it I didn't know Ruby maintained different scopes for that. We should probably do the same.

**Comment by straight-shoota** [comment]
Yes, it's actually two issues:
```cr
# No. 1
macro foo(arg)
end

foo bar = 1

bar # Syntax Error: read before assignment to local variable 'bar'
```
It should be `undefined local variable or method 'bar'`.

This example would work if `foo` was a method. But since it's a macro, the argument expression shouldn't be evaluated in the current scope but only when it is expanded in the macro body..

```cr
# No. 2
module Test
  foo = 1
end

p! foo # => 1
```
This should also error with `undefined local variable or method 'bar'`.

**Comment by asterite** [comment]
For number 1 the error is correct: method arguments are solved before the method call.

**Comment by asterite** [comment]
Ah, I misread 1. Maybe the error message could be improved, but it's still an error. The parser considers `bar` a variable, because if its previous assignment, so it can never be considered a method in the third line. The error message can't be "undefined local variable **or method**" because of that.

Maybe we can disallow assignments to variables inside method calls, except when they are inside macros. The parser then wouldn't treat subsequent uses of `bar` as a local variable. Assigning a variable inside a call would then be an error in the semantic phase.

Thoughts?

**Comment by straight-shoota** [comment]
To summarize, the parser recognizes a local var assignment and registers the variable in the current scope. It can't know that the call is a macro call and the argument expression will not be interpreted as an actual variable assignment. Thus the variable is only declared, not initialized.
Call resolution only happens later at the semantic phase and variables need to be known by then.

So the problem is that parsing the argument of a macro call has side effects which result in a syntax error later in the code. The compiler doesn't even get to the point to recognize `foo` as a macro call.

I don't know if there are similar expressions that have a side effect at the parsing stage and could trigger some weird errors as well. It might only be local variables, actually...

Possible solutions:

* Disallow assignments to local variables in method arguments. It's probably not that an important feature, and might even force to write cleaner code... but then, it's a bit weird when method calls like `foo.foo = 1` are allowed but local var assignments like `foo = 1` don't compiler. You mentioned this would raise in the semantic phase... wouldn't that be too late and the parser still error with `read before assignment`?
* Introduce some designation to differentiate macro calls from method calls. This has been suggested before, but it's certainly way to big a change for this little issue.
* Maybe it's possible to hold off the syntax error about uninitialized variables until macro calls are resolved. I doubt this is doable because local variables influence method resolution. Even if it could work, it would be pretty complicated.
* Given that this is a very rare issue and with No 2. changed, the impact is even less significant, it might just be fine to accept this error. The initial example would be fixed. And assignments as macro arguments are typically only used like this for `property` (etc.) declarations with default value. Using local variables in type scope is pretty rare, so the intersection of both, where this error can appear, should be negligible.
